File: linkedlist/doubly_ll_basics_strivers.java
class Node{
    int data;
    Node next;
    Node prev;

    Node(int data1){
        this.data = data1;
        this.next = null;
        this.prev = null;
    }
    Node(int data1, Node next1, Node prev1){
        this.data = data1;
        this.next = next1;
        this.prev = prev1;
    }
}

public class doubly_ll_basics_strivers {
    private static Node ConvertArrtoDLL(int[] arr){
        if (arr.length == 0) return null;
        Node head = new Node(arr[0]);
        Node tail = head;
        for(int i = 1; i < arr.length; i++){
            Node newNode = new Node(arr[i], null, tail);
            tail.next = newNode;
            tail = newNode;
        }
        return head;
    }
    private static void print(Node head){
        System.out.println("printing DLL");
        while(head != null){
            System.out.println(head.data);
            head = head.next;
        }
    }
    public static void main(String[] args){
        int[] arr = {1,2,3,4,5,6};
        Node head = ConvertArrtoDLL(arr);
        print(head);

    }

}

File: linkedlist/doubly_ll_basics_strivers.cpp
#include <bits/stdc++.h>
using namespace std;

class Node {
public:
    int data;
    Node* next;
    Node* prev;

    Node(int data) : data(data), next(nullptr), prev(nullptr) {}
    Node(int data, Node* next, Node* prev) : data(data), next(next), prev(prev) {}
};

Node* convertArrtoDLL(const vector<int>& arr) {
    if (arr.empty()) return nullptr;
    Node* head = new Node(arr[0]);
    Node* tail = head;
    for (size_t i = 1; i < arr.size(); ++i) {
        Node* newNode = new Node(arr[i]);
        tail->next = newNode;
        newNode->prev = tail;
        tail = newNode;
    }
    return head;
}

void printDLL(Node* head) {
    cout << "printing DLL\n";
    while (head != nullptr) {
        cout << head->data << " ";
        head = head->next;
    }
    cout << '\n';
}

int main() {
    vector<int> arr = {12, 5, 8, 7};
    Node* head = convertArrtoDLL(arr);
    printDLL(head);
    return 0;
}